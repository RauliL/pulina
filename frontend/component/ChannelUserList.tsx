import * as React from 'react';

export interface StateProps {
  users?: string[];
}

export type Props = StateProps;

const ChannelUserList: React.SFC<Props> = ({ users }) => (
  <ul className="list-unstyled" style={{ overflowY: 'auto' }}>
    {users && users.sort().map((user) => (
      <li className="p-2 border-bottom" key={user}>
        {user}
      </li>
    ))}
  </ul>
);

export default ChannelUserList;
