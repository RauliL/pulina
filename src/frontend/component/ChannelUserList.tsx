import React, { FunctionComponent } from "react";
import { Button } from "reactstrap";

export type StateProps = {
  channelName?: string;
  users: string[];
};

export type DispatchProps = {
  onToggleUserList: () => void;
};

export type Props = StateProps & DispatchProps;

const ChannelUserList: FunctionComponent<Props> = (props) => (
  <ul className="list-unstyled" style={{ overflowY: "auto" }}>
    <li className="d-lg-none navbar border-bottom">
      {props.channelName && (
        <span className="navbar-brand">Users on {props.channelName}</span>
      )}
      <div className="float-right">
        <Button outline={true} onClick={props.onToggleUserList}>
          <i className="fas fa-times" />
        </Button>
      </div>
    </li>
    {props.users.map((user) => (
      <li className="p-2 border-bottom" key={user}>
        {user}
      </li>
    ))}
  </ul>
);

export default ChannelUserList;
