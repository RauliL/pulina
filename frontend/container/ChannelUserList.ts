import { connect } from 'react-redux';

import { toggleUserList } from '../action';
import ChannelUserList, {
  DispatchProps,
  StateProps,
} from '../component/ChannelUserList';
import { PulinaDispatch, PulinaState } from '../types';
import { getCurrentChannel } from '../utils';

const mapStateToProps = (state: PulinaState): StateProps => {
  const currentChannel = getCurrentChannel(state);
  let channelName: string | undefined;
  let users: string[] | undefined;

  if (currentChannel) {
    channelName = currentChannel.name;
    users = currentChannel.users.sort();
  } else {
    users = [];
  }

  return { channelName, users };
};

const mapDispatchToProps = (dispatch: PulinaDispatch): DispatchProps => ({
  onToggleUserList() {
    dispatch(toggleUserList());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ChannelUserList);
