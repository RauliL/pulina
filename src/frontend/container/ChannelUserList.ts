import { connect } from "react-redux";

import { toggleUserList } from "../action";
import ChannelUserList, {
  DispatchProps,
  StateProps,
} from "../component/ChannelUserList";
import { Dispatch, State } from "../types/store";
import { getCurrentChannel } from "../utils";

const mapStateToProps = (state: State): StateProps => {
  const currentChannel = getCurrentChannel(state);
  let channelName: string | undefined;
  let users: string[] | undefined;

  if (currentChannel) {
    channelName = currentChannel.name;
    users = currentChannel.users.sort();
  } else {
    users = [];
  }

  return { channelName, users };
};

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
  onToggleUserList() {
    dispatch(toggleUserList());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ChannelUserList);
