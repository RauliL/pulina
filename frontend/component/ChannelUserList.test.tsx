import { mount, shallow } from 'enzyme';
import { defaultTo, noop } from 'lodash';
import * as React from 'react';
import { Button } from 'reactstrap';

import ChannelUserList, { Props } from './ChannelUserList';

describe('<ChannelUserList/>', () => {
  const renderComponent = (props: Partial<Props> = {}) => (
    <ChannelUserList
      channelName={props.channelName}
      users={defaultTo(props.users, [])}
      onToggleUserList={defaultTo(props.onToggleUserList, noop)}
    />
  );
  const mountComponent = (props: Partial<Props> = {}) => mount(
    renderComponent(props),
  );

  it('should render without throwing exception', () => {
    expect(shallow(renderComponent())).toMatchSnapshot();
  });

  it('should render <ul/>', () => {
    expect(mountComponent().find('ul')).toHaveLength(1);
  });

  it('should render each user with <li/>', () => {
    const component = mountComponent({
      users: [
        'test-1',
        'test-2',
        'test-3',
      ],
    });

    // It's actually 4 <li/>s because the toggle button is also rendered inside
    // a <li/>.
    expect(component.find('li')).toHaveLength(4);
  });

  it('should render a <Button/>', () => {
    expect(mountComponent().find(Button)).toHaveLength(1);
  });

  it('should invoke callback when <Button/> is being clicked', () => {
    const onToggleUserList = jest.fn();
    const component = mountComponent({ onToggleUserList });

    component.find(Button).simulate('click');
    expect(onToggleUserList).toBeCalled();
  });

  it('should optionally render channel name', () => {
    const component = mountComponent({ channelName: '#test-channel' });

    expect(component.find('li').at(0).text()).toBe('Users on #test-channel');
  });
});
