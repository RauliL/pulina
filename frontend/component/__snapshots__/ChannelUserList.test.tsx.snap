// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelUserList/> should render without throwing exception 1`] = `
<ul
  className="list-unstyled"
  style={
    Object {
      "overflowY": "auto",
    }
  }
>
  <li
    className="d-lg-none navbar border-bottom"
  >
    <div
      className="float-right"
    >
      <Button
        color="secondary"
        onClick={[Function]}
        outline={true}
        tag="button"
      >
        <i
          className="fas fa-times"
        />
      </Button>
    </div>
  </li>
</ul>
`;
